vignettes/s3-vector.Rmd

```diff
@@ -940,27 +931,223 @@ vec_arith.vctrs_meter.MISSING <- function(op, x, y, ...) {
 ```
 
-You also need to register the individual double-dispatch methods. Again, this is harder than it should be because roxygen's heuristics aren't quite right. That means you need to describe the `@method` explicitly:
+Now we finish the work around `vec_ptype2()` by providing all needed methods. Each method has both a `@method` and an `@export` directive. These are purely methods so they don't need to be documented in an `.Rd` file.
 
 ```{r}
-#' @method vec_cast.binned double
+#' @method vec_ptype2.vctrs_percent default
+#' @export                                 
+vec_ptype2.vctrs_percent.default <- function(x, y,
+                                             ...,
+                                             x_arg = "x", y_arg = "y") {
+  vec_default_ptype2(x, y, x_arg = x_arg, y_arg = y_arg)
+}
+
+#' @method vec_ptype2.vctrs_percent vctrs_percent
+#' @export                                       
+vec_ptype2.vctrs_percent.vctrs_percent <- function(x, y, ...) new_percent()
+
+#' @method vec_ptype2.vctrs_percent double       
+#' @export                                       
+vec_ptype2.vctrs_percent.double <- function(x, y, ...) double()
+
+#' @method vec_ptype2.double vctrs_percent
 #' @export
-vec_cast.binned.double <- function(x, y, ...) {
+vec_ptype2.double.vctrs_percent <- function(x, y, ...) double()
+```
+
+The exact same treatment applies to `vec_cast()`, which yields:
+
+```{r eval = FALSE}
+#' @method vec_cast pizza_percent
+#' @export vec_cast.pizza_percent
+#' @export                         
+#' @rdname pizza-vctrs             
+vec_cast.pizza_percent <- function(x, to, ...) {
+  UseMethod("vec_cast.pizza_percent")
 }
+#' @method vec_cast.pizza_percent default
+#' @export 
+vec_cast.pizza_percent.default <- function(x, to, ...) vec_default_cast(x, to)
+#' @method vec_cast.pizza_percent pizza_percent
+#' @export 
+vec_cast.pizza_percent.pizza_percent <- function(x, to, ...) x
+#' @method vec_cast.pizza_percent double
+#' @export
+vec_cast.pizza_percent.double <- function(x, to, ...) percent(x)
+#' @method vec_cast.double pizza_percent
+#' @export
+vec_cast.double.pizza_percent <- function(x, to, ...) vec_data(x)
 ```
 
-Hopefully future versions of roxygen will make these exports less painful.
+For simple casting and coercion functions, this is the recommended formatting. If everything fits on one line, there's no need to use `{}` or add blank lines between method definitions. This compact formatting reinforces the common purpose of these methods and their documentation.
+
+### Testing
+
+It's good practice to test your new class. Specific recommendations:
+
+* `R/percent.R` is the type of file where you really do want 100% test coverage.
+  You can use `devtools::test_file_coverage()` to check this.
+
+* Make sure to test behaviour with zero-length inputs and missing values.
+
+* Use `testthat::verify_output()` to add some 
+  [golden tests](https://ro-che.info/articles/2017-12-04-golden-tests) for 
```

Comment:
I think https://www.tidyverse.org/blog/2019/11/testthat-2-3-0/ might be a better resource. The golden test blog post doesn't mention the most important reason for using `verify_output()`, which is to not depend on the printing/deparsing/messaging behaviour of other packages.

Reply:
OK I added this to the current PR refining this contribution.